### PR TITLE
Add prefix-matching function to trie.py

### PR DESCRIPTION
**Pull request by tasbihaasim**
Added a new function match to trie.py. This function takes a prefix as an argument, traces that prefix in a trie and then iterates through all the succeeding letters in the sub-trie such that a list of words containing that prefix is returned. 

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: data_structures/trie/trie.py [hunk 1]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
             curr = curr.nodes[char]
         return curr.is_leaf
 
+    def match(self, prefix: str) -> [str]:
```
Comment (cclauss) [inline data_structures/trie/trie.py:49]: `match()` seems like a weak function name.  What about `words_starting_with()` or `get_words_starting_with()`?

Doctests?  How do we know this works?

File: data_structures/trie/trie.py [hunk 2]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
...
+        """
+        gets the list of all the words that
+        are followed by certain prefix.
+    
+        long-url: https://www.geeksforgeeks.org/prefix-matching-python-using-pytrie-module/
+        :param prefix: the prefix to be matched
+        :return: List of prefix-matched words.
+
+        """
+        pass
```
Comment (cclauss) [inline data_structures/trie/trie.py:59]: Helpful?

File: data_structures/trie/trie.py [hunk 3]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
...
+            :param t: self
+            :param prefix: prefix to be searched
+            :return: a new sub-trie
+            
+            '''
+            curr = t
+            for char in prefix:
+                if char not in curr.nodes:
+                    return False
+                    break
```
Comment (cclauss) [inline data_structures/trie/trie.py:76]: This `break` is unreachable code.  We have already returned False.

File: data_structures/trie/trie.py [hunk 4]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
...
+    
+        long-url: https://www.geeksforgeeks.org/prefix-matching-python-using-pytrie-module/
+        :param prefix: the prefix to be matched
+        :return: List of prefix-matched words.
+
+        """
+        pass
+
+
+        def search(t: TrieNode, prefix: str) -> TrieNode:
```
Comment (cclauss) [inline data_structures/trie/trie.py:62]: `search()` seems liek a weak function name.  What about `get_trie_for prefix()` or similar?

Doctests? How do we know this works?

File: data_structures/trie/trie.py [hunk 5]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
...
+            '''
+            curr = t
+            for char in prefix:
+                if char not in curr.nodes:
+                    return False
+                    break
+                curr = curr.nodes[char]
+            return curr
+
+        def dfs(root: TrieNode, s: str, prefix: str, lst):
```
Comment (cclauss) [inline data_structures/trie/trie.py:80]: `dfs()` -> `depth_first_search()` please.  Doctests?  How do we know this works?

File: data_structures/trie/trie.py [hunk 6]
```diff
@@ -46,6 +46,57 @@ def find(self, word: str) -> bool:
...
+            instead of True or False
+            :param t: self
+            :param prefix: prefix to be searched
+            :return: a new sub-trie
+            
+            '''
+            curr = t
+            for char in prefix:
+                if char not in curr.nodes:
+                    return False
```
Comment (cclauss) [inline data_structures/trie/trie.py:75]: The type hint says that we are going to return a TrieNode and False is not a TrieNode.

File: data_structures/trie/trie.py [hunk 7]
```diff
@@ -46,10 +46,10 @@ def find(self, word: str) -> bool:
             curr = curr.nodes[char]
         return curr.is_leaf
 
-    def match(self, prefix: str) -> [str]:
+    def get_words_starting_with(self, prefix: str) -> [str]:
         """
         gets the list of all the words that
```
Comment (cclauss) [inline data_structures/trie/trie.py:51]:
```suggestion
        return a list of all words that start with the given prefix
```

File: data_structures/trie/trie.py [hunk 8]
```diff
@@ -89,9 +92,7 @@ def depth_first_search(root: TrieNode, s: str, prefix: str, lst):
         if sub_trie==False:
             return False
         else:
-            lst=[]
-            l= depth_first_search(sub_trie, '', prefix, lst)
-            return l
+            print_words(sub_trie, prefix)
```
Comment (cclauss) [inline data_structures/trie/trie.py:95]: Algorithm functions should not print().  Please return the results to the caller.

File: data_structures/trie/trie.py [hunk 9]
```diff
@@ -89,9 +92,7 @@ def depth_first_search(root: TrieNode, s: str, prefix: str, lst):
         if sub_trie==False:
             return False
```
Comment (cclauss) [inline data_structures/trie/trie.py:93]: Rather than returning False, why not return an empty TrieNode?

**Comment by tasbihaasim** [comment]
@cclauss Hey can you look at my changes once again. The code is building and passing pytests but it is still failing Travis Cl and codespell tests. Thank you.

File: data_structures/trie/trie.py (hunk 8 above)
Comment (doudou-dounsm) [reply data_structures/trie/trie.py]: return后面还可以print()?

Comment (doudou-dounsm) [reply data_structures/trie/trie.py]: The idea might be to add two patterns,
But it doesn't seem necessary.

**Comment by cclauss** [comment]
Codespell says ./data_structures/trie/trie.py:84: resturns  ==> returns

Travis says >>> get_words_starting_with("a")
NameError("name 'get_words_starting_with' is not defined")

File: data_structures/trie/trie.py [hunk 10]
```diff
@@ -55,7 +55,7 @@ def get_words_starting_with(self, prefix: str) -> [str]:
         :param prefix: the prefix to be matched
```
Comment (cclauss) [inline data_structures/trie/trie.py:55]: # List[str]

File: data_structures/trie/trie.py [hunk 11]
```diff
@@ -46,6 +46,65 @@ def find(self, word: str) -> bool:
...
+            for char in prefix:
+                if char not in curr.nodes:
+                    return False
+                    
+                curr = curr.nodes[char]
+            return curr
+
+        def depth_first_search(node: TrieNode, word: str, lst):
+            """
+            resturns a list of all the nodes in a trie
```
Comment (mrmaxguns) [inline data_structures/trie/trie.py:84]: Line 84 is causing codespell tests to fail. Easy typo `resturns` to `returns`

**Comment by cclauss** [comment]
On line 38 can you please rename `find()` to `__contains__()` so we can use `in` ?
https://docs.python.org/3/reference/datamodel.html#object.__contains__

**Pull request closed without merging**